p11-kit/proxy.c

```diff
@@ -604,7 +604,40 @@ proxy_C_WaitForSlotEvent (CK_X_FUNCTION_LIST *self,
                           CK_SLOT_ID_PTR slot,
                           CK_VOID_PTR reserved)
 {
-	return CKR_FUNCTION_NOT_SUPPORTED;
+	State *state = (State *)self;
+	Proxy *py = state->px;
+	CK_FUNCTION_LIST_PTR *f;
+	CK_FUNCTION_LIST_PTR funcs;
+	CK_SLOT_ID real_slot;
+	unsigned int i;
+	CK_RV rv = CKR_NO_EVENT;
+
+	/* Only the non-blocking case is supported. */
+	if ((flags & CKF_DONT_BLOCK) == 0)
+		return CKR_FUNCTION_NOT_SUPPORTED;
+
+	p11_lock ();
+
+	for (f = py->inited; *f; ++f) {
+		funcs = *f;
+		assert (funcs != NULL);
+
+		rv = (funcs->C_WaitForSlotEvent) (flags, &real_slot, reserved);
+		if (rv == CKR_NO_EVENT)
+			continue;
```

Comment:
This looks like a busy waiting loop, would it make sense to add some sleeping with a exponential backoff ?

Reply:
I think it should be okay, because the loop is calling `C_WaitForSlotEvent` for different modules one after another.

Reply:
Ah sorry,  imissed that this cycles through all modules only once. Nevermind